File: pickerkit/src/main/java/com/dewarder/pickerkit/provider/PickerMediaVideoWrapper.java
package com.dewarder.pickerkit.provider;

import com.annimon.stream.Stream;
import com.dewarder.pickerkit.model.PickerMedia;
import com.dewarder.pickerkit.model.PickerVideo;

import java.util.Collection;

class PickerMediaVideoWrapper implements PickerDataProvider<PickerMedia> {

    private final PickerDataProvider<PickerVideo> delegate;

    private PickerMediaVideoWrapper(PickerDataProvider<PickerVideo> delegate) {
        this.delegate = delegate;
    }

    public static PickerDataProvider<PickerMedia> wrap(PickerDataProvider<PickerVideo> provider) {
        return new PickerMediaVideoWrapper(provider);
    }

    @Override
    public void request(Callback<PickerMedia> callback) {
        delegate.request(new Callback<PickerVideo>() {
            @Override
            public void onNext(Collection<PickerVideo> data) {
                callback.onNext(
                        Stream.of(data)
                                .map(PickerVideo::getSource)
                                .map(PickerMedia.Companion::video)
                                .toList());
            }

            @Override
            public void onComplete() {
                callback.onComplete();
            }

            @Override
            public void onError(Throwable throwable) {
                callback.onError(throwable);
            }
        });
    }
}

File: pickerkit/src/main/java/com/dewarder/pickerkit/provider/PickerMediaVideoWrapper.kt
package com.dewarder.pickerkit.provider

import com.dewarder.pickerkit.gallery.model.PickerMedia
import com.dewarder.pickerkit.gallery.model.PickerVideo

internal class PickerMediaVideoWrapper private constructor(
        private val delegate: PickerDataProvider<PickerVideo>
) : PickerDataProvider<PickerMedia> {

    override fun request(callback: PickerDataProvider.Callback<PickerMedia>) {
        delegate.request(object : PickerDataProvider.Callback<PickerVideo> {
            override fun onNext(data: Collection<PickerVideo>) {
                callback.onNext(
                        data.map { it.source }.map { PickerMedia.video(it) }
                )
            }

            override fun onComplete() {
                callback.onComplete()
            }

            override fun onError(throwable: Throwable) {
                callback.onError(throwable)
            }
        })
    }

    companion object {

        fun wrap(provider: PickerDataProvider<PickerVideo>): PickerDataProvider<PickerMedia> =
                PickerMediaVideoWrapper(provider)
    }
}
